src/hotspot/share/runtime/flags/jvmFlagAccess.cpp

```diff
@@ -0,0 +1,397 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "jfr/jfrEvents.hpp"
+#include "memory/allocation.hpp"
+#include "runtime/flags/jvmFlag.hpp"
+#include "runtime/flags/jvmFlagAccess.hpp"
+#include "runtime/flags/jvmFlagLimit.hpp"
+#include "runtime/flags/jvmFlagConstraintsRuntime.hpp"
+#include "utilities/macros.hpp"
+#include "utilities/ostream.hpp"
+
+template<typename T, typename EVENT>
+static void trace_flag_changed(JVMFlag* flag, const T old_value, const T new_value, const JVMFlag::Flags origin) {
+  EVENT e;
+  e.set_name(flag->name());
+  e.set_oldValue(old_value);
+  e.set_newValue(new_value);
+  e.set_origin(origin);
+  e.commit();
+}
+
+class FlagAccessImpl {
+public:
+  JVMFlag::Error set(JVMFlag* flag, void* value, JVMFlag::Flags origin) const {
+    return set_impl(flag, value, origin);
+  }
+
+  virtual JVMFlag::Error set_impl(JVMFlag* flag, void* value, JVMFlag::Flags origin) const = 0;
+  virtual JVMFlag::Error check_range(const JVMFlag* flag, bool verbose) const { return JVMFlag::SUCCESS; }
+  virtual void print_range(outputStream* st, const JVMFlagLimit* range) const { ShouldNotReachHere(); }
+  virtual void print_default_range(outputStream* st) const { ShouldNotReachHere(); }
+  virtual JVMFlag::Error check_constraint(const JVMFlag* flag, void * func, bool verbose) const  { return JVMFlag::SUCCESS; }
+};
+
+template <typename T, int type_enum, typename EVENT>
+class TypedFlagAccessImpl : public FlagAccessImpl {
+
+public:
+  JVMFlag::Error check_constraint_and_set(JVMFlag* flag, void* value_addr, JVMFlag::Flags origin, bool verbose) const {
+    T value = *((T*)value_addr);
+    const JVMTypedFlagLimit<T>* constraint = (const JVMTypedFlagLimit<T>*)JVMFlagLimit::get_constraint(flag);
+    if (constraint != NULL && constraint->phase() <= static_cast<int>(JVMFlagLimit::validating_phase())) {
+      JVMFlag::Error err = typed_check_constraint(constraint->constraint_func(), value, verbose);
+      if (err != JVMFlag::SUCCESS) {
+        return err;
+      }
+    }
+
+    T old_value = flag->read<T, type_enum>();
+    trace_flag_changed<T, EVENT>(flag, old_value, value, origin);
+    flag->write<T, type_enum>(value);
+    *((T*)value_addr) = old_value;
+    flag->set_origin(origin);
+
+    return JVMFlag::SUCCESS;
+  }
+
+  JVMFlag::Error check_constraint(const JVMFlag* flag, void * func, bool verbose) const  {
+    return typed_check_constraint(func, flag->read<T, type_enum>(), verbose);
+  }
+
+  virtual JVMFlag::Error typed_check_constraint(void * func, T value, bool verbose) const = 0;
+};
+
+class FlagAccessImpl_bool : public TypedFlagAccessImpl<JVM_FLAG_TYPE(bool), EventBooleanFlagChanged> {
+public:
+  JVMFlag::Error set_impl(JVMFlag* flag, void* value_addr, JVMFlag::Flags origin) const {
+    bool verbose = !JVMFlagLimit::validated_after_ergo();
```

Comment:
Why only verbose when not AfterErgo? Maybe add comment explaining that.

Reply:
If I understand the old code correctly, we print out error messages only when parsing the command-line given by the user. When we later override values that are set programmatically by the FLAG_SET_XXX() macros, we don't print out error messages (or else the user will be confused). I will rewrite the code to make it clear, and add comments.